pkgpanda/actions.py

```diff
@@ -0,0 +1,293 @@
+
+    # Validate the package id
+    PackageId(pkg_id)
+
+    def fetch(_, target):
+        extract_tarball(package_filename, target)
+
+    repository.add(fetch, pkg_id)
+
+
+def remove_packages(install, repository, package_ids):
+    """Remove packages from the local repository.
+
+    Errors if any packages in package_ids are activated in install.
+
+    install: pkgpanda.Install
+    repository: pkgpanda.Repository
+    package_ids: sequence of package IDs to remove from repository
+
+    """
+    assert isinstance(package_ids, collections.Sequence)
+
+    # Make sure none of the packages are active
+    active_packages = install.get_active()
+    active = active_packages.intersection(set(package_ids))
+    if len(active) > 0:
+        print("Refusing to remove active packages {0}".format(" ".join(sorted(list(active)))))
+        sys.exit(1)
+
+    for pkg_id in package_ids:
+        sys.stdout.write("\rRemoving: {0}".format(pkg_id))
+        sys.stdout.flush()
+        try:
+            # Validate package id, that package is installed.
+            PackageId(pkg_id)
+            repository.remove(pkg_id)
+        except ValidationError:
+            print("\nInvalid package id {0}".format(pkg_id))
+            sys.exit(1)
+        except OSError as ex:
+            print("\nError removing package {0}".format(pkg_id))
+            print(ex)
+            sys.exit(1)
+        sys.stdout.write("\rRemoved: {0}\n".format(pkg_id))
+        sys.stdout.flush()
+
+
+def setup(install, repository):
+    """Set up a fresh install of DC/OS.
+
+    install: pkgpanda.Install
+    repository: pkgpanda.Repository
+
+    """
+    # Check for /opt/mesosphere/bootstrap. If not exists, download everything
+    # and install /etc/systemd/system/mutli-user.target/dcos.target
+    bootstrap_path = os.path.join(install.root, "bootstrap")
+    if os.path.exists(bootstrap_path):
+        # Write, enable /etc/systemd/system/dcos.target for next boot.
+        dcos_target_dir = os.path.dirname(install.systemd_dir)
+        try:
+            os.makedirs(dcos_target_dir)
+        except FileExistsError:
+            pass
+
+        write_string(os.path.join(dcos_target_dir, "dcos.target"),
+                     DCOS_TARGET_CONTENTS)
+        _do_bootstrap(install, repository)
+        # Enable dcos.target only after we have populated it to prevent starting
+        # up stuff inside of it before we activate the new set of packages.
+        if install.manage_systemd:
+            _start_dcos_target(block_systemd=False)
+        os.remove(bootstrap_path)
+
+    # Check for /opt/mesosphere/install_progress. If found, recover the partial
+    # update.
+    if os.path.exists("/opt/mesosphere/install_progress"):
+        took_action, msg = install.recover_swap_active()
+        if not took_action:
+            print("No recovery performed: {}".format(msg))
+
+
+def _start_dcos_target(block_systemd):
+    no_block = [] if block_systemd else ["--no-block"]
```

Comment:
This is now in its own commit: https://github.com/dcos/dcos/pull/104/commits/22908f3ed1a8f4d1153bc9df30e5a05d3e5abfdd